### PR TITLE
Normalize Linux and Windows X.509 Revocation Statuses

### LINKED ISSUE #41475: X509Chain.Build failed on Linux(netcore3.0) with OCSP configured
Encounter an issue on Linux when trying to call X509Chain.Build method on a leaf certificate. The CA has configured OCSP but not CRL.

The ChainStatus has "unable to get certificate CRL". 

According to https://github.com/dotnet/corefx/pull/35367#issuecomment-535727309, CRL should not be checked if OCSP responds correctly.

Here are the repro steps:
    
1.	Download the netcoreapp3.0.zip
[netcoreapp3.0.zip](https://github.com/dotnet/corefx/files/3677735/netcoreapp3.0.zip)
 to your machine. 

2.	Open a session(naming as session 1) on Linux, copy the netcoreapp3.0 folder to the Linux machine.

3.	Run command:  dotnet ./netcoreapp3.0/ConsoleApp3.dll

3 certificates files will be generated under netcoreapp3.0 folder.  They’re 0.cer (leaf), 1.cer(intermediate), 2.cer(root)
  
Simulated CA servers will start running. (No CRL respond is configured, but OCSP respond is configured )

There will be following messages displayed:

The responds of simulated OCSP server to each requests will be displayed , when trying to run X509Chain.Build on the leaf certificate(0.cer)
(Every OCSP requests will have respond info(nonce & certificate status) displayed on session 1
```
CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (0f813155-18a9-4748-affd-dce5493ca3ab)
1 OCSP respond(s) is sent..
   nonce is #0410549e5736285ac06ad3f6d992123aff06
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (0f813155-18a9-4748-affd-dce5493ca3ab)
1 OCSP respond(s) is sent..
   nonce is #04109e605bd153602b04981ea21592d70ea1
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (0f813155-18a9-4748-affd-dce5493ca3ab)
1 OCSP respond(s) is sent..
   nonce is #0410fae8e895c936da8060ebc7ec62057f12
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (0f813155-18a9-4748-affd-dce5493ca3ab)
1 OCSP respond(s) is sent..
   nonce is #04109d32979804212d5c36112e0eee6edc5e
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Root Certificate Authority (7985adab-a107-415f-b2e8-81f1b161d709)
1 OCSP respond(s) is sent..
   nonce is #04101d22bbc6a0ffd7a85873fdf46cbb5814
   certificateStatus : Good
```
The results of chian.Build (go through ChainElement  of the chain.ChainElements and prints all chainElementStatus)
```
Revocation flag:  ExcludeRoot
chain building results are as following:
certificate.Subject is : CN=NuGet Test Timestamp Service (26558ee9-d5af-4ef8-87b5-3764eb322e94), O=NuGet, L=Redmond, S=WA, C=US
certificate.FriendlyName is :
=================   The (0) th certificate is   :=================
File path:  /home/henli/repos/NuGetTest/netcoreapp3.0/0.cer
chainElement.Certificate.Subject : (CN=NuGet Test Timestamp Service (26558ee9-d5af-4ef8-87b5-3764eb322e94), O=NuGet, L=Redmond, S=WA, C=US)
chainElement.Certificate.Thumbprint : (8906F366C606DCFFC348B71DD11EB4C093AE5177)
chainElement.Certificate.isvalid : (False)
    --------   The chainElementStatus are : -------
  status : (RevocationStatusUnknown)
  info: (unable to get certificate CRL)
=================   The (1) th certificate is   :=================
File path:  /home/henli/repos/NuGetTest/netcoreapp3.0/1.cer
chainElement.Certificate.Subject : (CN=NuGet Test Intermediate Certificate Authority (0f813155-18a9-4748-affd-dce5493ca3ab), O=NuGet, L=Redmond, S=WA, C=US)
chainElement.Certificate.Thumbprint : (D14A10A221063426451026C156E7C9AB46970554)
chainElement.Certificate.isvalid : (False)
    --------   The chainElementStatus are : -------
  status : (RevocationStatusUnknown)
  info: (unable to get certificate CRL)
=================   The (2) th certificate is   :=================
File path:  /home/henli/repos/NuGetTest/netcoreapp3.0/2.cer
chainElement.Certificate.Subject : (CN=NuGet Test Root Certificate Authority (7985adab-a107-415f-b2e8-81f1b161d709), O=NuGet, L=Redmond, S=WA, C=US)
chainElement.Certificate.Thumbprint : (8E95B08D328C0414890E244DFCD2FB5A09EF81F4)
chainElement.Certificate.isvalid : (False)
    --------   The chainElementStatus are : -------
  status : (RevocationStatusUnknown)
  info: (unable to get certificate CRL)
```
4.	When “The CA servers are running, you may start another session to test ..” is displayed, duplicate a session on Linux (naming as session 2).

5.	On session 2, open netcoreapp3.0 folder, run the following commands to generate .pem files for certificate files generated in step3:

  openssl x509 -in ./0.cer -inform der -outform pem -out ./0.pem

  openssl x509 -in ./1.cer -inform der -outform pem -out ./1.pem

   openssl x509 -in ./2.cer -inform der -outform pem -out ./2.pem

6.	On session 2, run command “openssl x509 -in ./0.pem -text -noout” , find the  OCSP URL( the line starts with OCSP – URI)

7.	On session 2, run command ```openssl ocsp -issuer ./1.pem -cert ./0.pem -text -url ``` + the URL we  got from step 6

8.	Compare the outputs on session 1 and session 2, the nonce from session 1(OCSP respond) and the nonce from session 2(OCSP request) should be the same.  And the openssl verify results from session 2 is good.
```
Response verify OK
./0.pem: good
```

So step 8 shows that the CA server has configured OCSP respond with correct nonce and status. 
But the step 3 shows that the chain.Build failed for unable to get certificate CRL.

Above are the repro steps on Linux.

If we repro on windows(netcoreapp3.0), we have following different results:
Compared to step 3 on Linux, Windows has a null nonce in OCSP respond:

```
CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Root Certificate Authority (86b412bb-d4bf-40bd-a544-263ae3ed26be)
1 OCSP respond(s) is sent..
   nonce is null
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (9c6012a9-2ce2-4e07-b1a2-789dc2f2f219)
1 OCSP respond(s) is sent..
   nonce is null
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (9c6012a9-2ce2-4e07-b1a2-789dc2f2f219)
1 OCSP respond(s) is sent..
   nonce is null
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Root Certificate Authority (86b412bb-d4bf-40bd-a544-263ae3ed26be)
1 OCSP respond(s) is sent..
   nonce is null
   certificateStatus : Good

CertificateAuthority is : C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Intermediate Certificate Authority (9c6012a9-2ce2-4e07-b1a2-789dc2f2f219)
1 OCSP respond(s) is sent..
   nonce is null
   certificateStatus : Good
```

Windows has chain.Build running successfully: (status are empty)

```
Revocation flag:  ExcludeRoot
chain building results are as following:
certificate.Subject is : CN=NuGet Test Timestamp Service (9cef8b66-4871-46ae-b6ee-a411937ebecf), O=NuGet, L=Redmond, S=WA, C=US
certificate.FriendlyName is :
=================   The (0) th certificate is   :=================
File path:  C:\repos\NuGet.Client\test\NuGet.Core.FuncTests\ConsoleApp3\bin\Debug\netcoreapp3.0\0.cer
chainElement.Certificate.Subject : (CN=NuGet Test Timestamp Service (9cef8b66-4871-46ae-b6ee-a411937ebecf), O=NuGet, L=Redmond, S=WA, C=US)
chainElement.Certificate.Thumbprint : (13C3BFE21E62DF8D916B56BE8EE026CA35F98C92)
chainElement.Certificate.isvalid : (True)
    --------   The chainElementStatus are : -------
=================   The (1) th certificate is   :=================
File path:  C:\repos\NuGet.Client\test\NuGet.Core.FuncTests\ConsoleApp3\bin\Debug\netcoreapp3.0\1.cer
chainElement.Certificate.Subject : (CN=NuGet Test Intermediate Certificate Authority (9c6012a9-2ce2-4e07-b1a2-789dc2f2f219), O=NuGet, L=Redmond, S=WA, C=US)
chainElement.Certificate.Thumbprint : (F909EF774F675E0C262B5EF4D62830BA4C36EF8E)
chainElement.Certificate.isvalid : (True)
    --------   The chainElementStatus are : -------
=================   The (2) th certificate is   :=================
File path:  C:\repos\NuGet.Client\test\NuGet.Core.FuncTests\ConsoleApp3\bin\Debug\netcoreapp3.0\2.cer
chainElement.Certificate.Subject : (CN=NuGet Test Root Certificate Authority (86b412bb-d4bf-40bd-a544-263ae3ed26be), O=NuGet, L=Redmond, S=WA, C=US)
chainElement.Certificate.Thumbprint : (C3DBDC6825FB9C2348F0441494463073B55EAE4E)
chainElement.Certificate.isvalid : (True)
--------   The chainElementStatus are : -------
```

We also tried to run “openssl verify” command on Linux to verify certificate as following:
1.	On Linux session 2, run following command to  concatenate a chain:
cat ./1.pem > ./chain.pem
cat ./2.pem >> ./chain.pem

2.	Run “openssl verify -verbose -CAfile ./chain.pem ./0.pem”
```Result is :   ./0.pem: OK```

### LINKED ISSUE #41738: [Linux] X.509 chain building:  EndCertificateOnly does not work when end entity is covered only by OCSP
Found while investigating https://github.com/dotnet/corefx/issues/41475.

CC @bartonjs

### PR DESCRIPTION
**Pull request by bartonjs**
* Add a bunch of tests for X.509 Revocation using an HttpListener-based responder
* Add Linux support for OCSP at multiple levels (not just the leaf)
* Fix up some inconsistencies in chain element statuses

Fixes #41738.
Fixes #41475.

After CR signoff I'll flip the entire DynamicRevocationTests class to OuterLoop, and possibly look into finding out ways to get some/better parallelization out of it.  (I tried using a rotating pool of RSA keys for the test, but that didn't measurably move the needle in the runs on my machine... it's dominated by serial execution of serial loopback HTTP)

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs [hunk 1]
```diff
@@ -30,13 +32,16 @@ internal sealed class OpenSslX509ChainProcessor : IChainPal
         private static readonly CachedDirectoryStoreProvider s_userPersonalStore =
             new CachedDirectoryStoreProvider(X509Store.MyStoreName);
 
+        private static readonly ConcurrentDictionary<Interop.Crypto.X509VerifyStatusCode, string> s_errorStrings =
+            new ConcurrentDictionary<Interop.Crypto.X509VerifyStatusCode, string>();
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Can you add a comment about why it's worth having this cache for error strings?  Presumably it's because we retrieve these in cases where they're not actually considered errors?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs [hunk 2]
```diff
@@ -352,69 +370,243 @@ internal void CommitToChain()
...
+                        encounteredRevocation = false;
+                    }
+                    else if (refErrors.HasRevocationUnknown() && revocationFlag != X509RevocationFlag.ExcludeRoot)
+                    {
+                        // If the chain size is 1 we need to copy the root cert into untrusted so
+                        // OCSP_basic_verify can find it.
+                        if (chainSize == 1)
+                        {
+                            using (SafeSharedX509StackHandle untrusted =
+                                Interop.Crypto.X509StoreCtxGetSharedUntrusted(_storeCtx))
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Nit: can you move this up to the end of the previous line?  I realize that makes it a little long, but these nested usings at the same level then look strange with it wrapped in this way.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs [hunk 3]
```diff
@@ -352,69 +370,243 @@ internal void CommitToChain()
...
+                // Reset to a WorkingChain that won't fail.
+                extraDispose = workingChain;
+                workingChain = new WorkingChain(abortOnSignatureError: false);
+                workingCallback = workingChain.VerifyCallback;
+                Interop.Crypto.X509StoreCtxSetVerifyCallback(_storeCtx, workingCallback);
+
+                verify = Interop.Crypto.X509VerifyCert(_storeCtx);
+            }
+
+            GC.KeepAlive(workingCallback);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Why is this necessary?  We're referencing workingChain below, and workingCallback is the same as workingChain.VerifyCallback, right?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs [hunk 4]
```diff
@@ -1083,6 +1377,30 @@ public Enumerator GetEnumerator()
...
+
+                foreach (var code in this)
+                {
+                    if (delim)
+                        builder.Append(" | ");
+
+                    builder.Append(code);
+                    delim = true;
+
+                }
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Nit: you could simplify this if you wanted:
```C#
string delim = "";
foreach (X509VerifyStatusCode code in this)
{
    builder.Append(code);
    delim = " | ";
}
```

**Review by stephentoub — APPROVED** [review]

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Yeah, particularly if a chain is built with suppression flags, then things like "unable to get certificate CRL" will get copied over and over again.  In the Windows version we preload all of the strings (we'll ever care about) during static field init.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs (hunk 3 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: `workingChain.VerifyCallback` is a method(group), `workingCallback` is an instance-bound delegate that was passed as a native callback.

Comment (stephentoub) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Ah, I thought it was a field / field-backed-property.  Nevermind.

Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Added a comment anyways :)

**Comment by heng-liu** [comment]
Hi @bartonjs , may I know if on macOS, the X.509 Revocation using an HttpListener-based responder could work correctly?
So we have a root cert -> intermediate cert -> leaf cert, 
then we have a simulated CA, revoking the leaf cert (by responding OCSP info, no CRL is involved) . 
On Windows and Linux, the X509Chain.Build failed for certificate revoked.
But on MacOS(10.14,10.15), the X509Chain.Build succeeded, there is no complain about the revoked certificate.

**Comment by bartonjs** [comment]
Is it never asking, or just that it's caching responses with no expiration?  It's pretty much a black box, we say "do revocation", and I've generally see it work for already-revoked certs; but (if I recall correctly) I wasn't having success with the tests doing dynamic state on macOS because of something like an apparent system cache.

**Comment by heng-liu** [comment]
Thanks @bartonjs, I will check if it's using caching responses or not.
In our test, we first wait until the the "nextUpdate" expire, then do the revoke. I'm not sure if doing in this way can avoid using the system cache. It works on Windows and Linux, but not MacOS.

**Comment by heng-liu** [comment]
Hi @zkat, FYI.
